File: host/src/main/java/com/reservation/host/auth/token/RequestContext.java
package com.reservation.host.auth.token;

import static com.reservation.host.auth.controller.AuthController.*;
import static org.springframework.context.annotation.ScopedProxyMode.*;

import java.util.Arrays;

import org.springframework.stereotype.Component;
import org.springframework.web.context.annotation.RequestScope;

import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;

@Component
@RequestScope(proxyMode = TARGET_CLASS)
@RequiredArgsConstructor
public class RequestContext {
	private final HttpServletRequest request;

	public String getRefreshToken() {
		return getCookie(REFRESH_COOKIE_NAME);
	}

	public String getCookie(String name) {
		if (request.getCookies() == null) {
			return null;
		}

		return Arrays.stream(request.getCookies())
			.filter(c -> c.getName().equals(name))
			.map(Cookie::getValue)
			.findFirst()
			.orElse(null);
	}
}

File: identity-service/src/main/kotlin/com/msa/identityservice/auth/token/RequestContext.kt
package com.msa.identityservice.auth.token

import com.msa.identityservice.auth.consts.AuthConstants.AUTH_HEADER_NAME
import com.msa.identityservice.auth.consts.AuthConstants.AUTH_HEADER_PREFIX
import com.msa.identityservice.auth.consts.AuthConstants.CONTEXT_NOT_FOUND_MESSAGE
import com.msa.identityservice.auth.consts.AuthConstants.REFRESH_COOKIE_NAME
import com.msa.identityservice.exception.BusinessErrorCode
import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletRequest
import org.springframework.context.annotation.ScopedProxyMode
import org.springframework.stereotype.Component
import org.springframework.web.context.annotation.RequestScope
import java.util.*


@Component
@RequestScope(proxyMode = ScopedProxyMode.TARGET_CLASS)
class RequestContext(
    private val request: HttpServletRequest
) {

    fun getRefreshToken(): String {
        if (request.cookies == null) {
            throw BusinessErrorCode.UNAUTHORIZED.exception(CONTEXT_NOT_FOUND_MESSAGE)
        }

        return Arrays.stream(request.cookies)
            .filter { c -> c.name.equals(REFRESH_COOKIE_NAME) }
            .map { obj: Cookie -> obj.value }
            .findFirst()
            .orElseThrow { BusinessErrorCode.UNAUTHORIZED.exception(CONTEXT_NOT_FOUND_MESSAGE) }
    }

    fun getAccessToken(): String {
        val authHeader = request.getHeader(AUTH_HEADER_NAME)
        if (authHeader == null || !authHeader.startsWith(AUTH_HEADER_PREFIX)) {
            throw BusinessErrorCode.UNAUTHORIZED.exception(CONTEXT_NOT_FOUND_MESSAGE)
        }

        return authHeader.substring(AUTH_HEADER_PREFIX.length)
    }

}
